tests/runner-tests/file-data-sources/src/mapping.ts

```diff
@@ -4,130 +4,148 @@ import {
+    entity.save();
+    let context = new DataSourceContext();
+    context.setString("command", CREATE_FOO);
+    dataSource.createWithContext("File", [event.params.data], context);
+  }
 
-export function handleFile1(data: Bytes): void {
-  let entity = new IpfsFile1(dataSource.stringParam());
-  entity.content = data.toString();
-  entity.save();
-}
+  if (event.params.testCommand == ACCESS_FILE_ENTITY_THROUGH_DERIVED_FIELD) {
+    let entity = Foo.load(event.params.data);
+    if (entity == null) {
+      log.info("Entity not found", []);
+    } else {
+      log.info("Accessing file entity through derived field", []);
+      let fileEntity = entity.ipfs.load();
 
-// Used to test spawning a file data source from another file data source handler.
-// This function spawns a file data source that will be handled by `handleSpawnedTest`.
-export function spawnTestHandler(data: Bytes): void {
-  let context = new DataSourceContext();
-  context.setString("file", "fromSpawnTestHandler");
-  let command = dataSource.context().getString("command");
-  if (command == "spawnOffChainHandlerTest") {
-    dataSource.createWithContext("OffChainDataSource", [KNOWN_CID], context);
-  } else if (command == "spawnOnChainHandlerTest") {
-    dataSource.createWithContext("OnChainDataSource", [KNOWN_CID], context);
+      assert(fileEntity.length == 0, "Expected exactly one file entity");
+    }
   }
 }
 
-// This is the handler for the data source spawned by `spawnOffChainHandlerTest`.
-export function handleSpawnedTest(data: Bytes): void {
-  let entity = new SpawnTestEntity(dataSource.stringParam());
-  let context = dataSource.context().getString("file");
-  entity.content = data.toString();
-  entity.context = context;
-  entity.save();
+export function handleFile(data: Bytes): void {
+  log.info("handleFile {}", [dataSource.stringParam()]);
+  let context = dataSource.context();
+
+  if (context.isSet("command")) {
+    let contextCommand = context.getString("command");
+
+    if (contextCommand == SPAWN_FDS_FROM_OFFCHAIN_HANDLER) {
+      let hash = context.getString("hash");
+      log.info("Creating file data source from handleFile: {}", [hash]);
+      dataSource.createWithContext("File", [hash], new DataSourceContext());
+    }
+
+    if (contextCommand == ONCHAIN_FROM_OFFCHAIN) {
+      log.info("Creating onchain data source from offchain handler", []);
+      let address = context.getString("address");
+      dataSource.create("OnChainDataSource", [address]);
+    }
+
+    if (contextCommand == CREATE_UNDEFINED_ENTITY) {
+      log.info("Creating undefined entity", []);
+      let entity = new Foo(dataSource.stringParam());
+      entity.save();
+    }
+
+    if (contextCommand == CREATE_FOO) {
+      log.info("Creating FileEntity with relation to Foo", []);
+      let entity = new FileEntity(dataSource.stringParam());
+      entity.foo = dataSource.stringParam();
+      entity.content = data.toString();
+      entity.save();
+    }
+  } else {
```

Comment:
Can remove a nesting level by doing this block if not command and returning. Then you know all the rest are commands. as zoran said it's prolly more of a switch case, mostly for making it a bit easier to maintain and read if you think it's worth. But also none of the ifs return when match so that could create some weird test situations :laughing:

Reply:
Agreed, making the change.